### PR TITLE
feat: add token auth, configurable CORS for HTTP filters, CI/release fixes

### PR DESCRIPTION
**Pull request by lucasmundim**
## Summary

- Token-based authentication and configurable CORS for webvis and REST filters via a shared `http_security` module, opt-in via config params and `FILTER_*` env vars
- CI/release improvements: shared security scan workflow, all GitHub Actions bumped to latest, dependency CVE fixes
- Fix Docker image build race condition by polling PyPI before starting builds

## Configuration

```yaml
filters:
  - name: webvis
    config:
      auth_token: "my-secret-token"
      cors_origins: "https://portal.plainsight.tech"
```

Or via env vars:
```bash
FILTER_AUTH_TOKEN=my-secret-token
FILTER_CORS_ORIGINS=https://portal.plainsight.tech
```

## Files

| File | Change |
|------|--------|
| `openfilter/filter_runtime/filters/http_security.py` | New: shared auth + CORS module |
| `openfilter/filter_runtime/filters/webvis.py` | Use `configure_http_security()`, add config params |
| `openfilter/filter_runtime/filters/rest.py` | Use `configure_http_security()`, strip auth credentials from frame data |
| `tests/test_http_security.py` | 42 unit tests for auth + CORS |
| `tests/test_filter_webvis.py` | 18 integration tests for webvis config/auth/CORS |
| `tests/test_filter_rest.py` | 18 integration tests for REST config/auth/CORS/credential stripping |
| `.github/workflows/create-release.yaml` | Add wait-for-pypi step, bump all actions |
| `.github/workflows/security-scan.yaml` | Use shared workflow |
| `.github/workflows/ci.yaml` | Bump all actions |
| `VERSION` | v0.1.29 → v0.1.30 |
| `RELEASE.md` | v0.1.30 release notes |

## Test plan

- [x] 76 tests passing across all three test files
- [x] Existing webvis tests pass (4/4)
- [x] CI green

### CODE REVIEW

File: openfilter/filter_runtime/filters/http_security.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            if not request_token:
+                auth_header = request.headers.get('authorization', '')
+                if auth_header.lower().startswith('bearer '):
+                    request_token = auth_header[7:].strip()
+
+            if request_token != token:
+                return JSONResponse(
+                    status_code=401,
+                    content={'detail': 'Unauthorized'},
+                )
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py]: Fixed — now using `secrets.compare_digest()` for constant-time comparison. See 6e5350d.

File: openfilter/filter_runtime/filters/http_security.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    from starlette.middleware.base import BaseHTTPMiddleware
+    from starlette.requests import Request
+    from starlette.responses import JSONResponse
+
+    class TokenAuthMiddleware(BaseHTTPMiddleware):
+        async def dispatch(self, request: Request, call_next):
+            # Skip auth for CORS preflight requests
+            if request.method == 'OPTIONS':
+                return await call_next(request)
+
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py]: Good catch. Updated to only bypass auth when the request has both `Origin` and `Access-Control-Request-Method` headers (i.e. an actual CORS preflight). Added a test for plain OPTIONS requiring auth. See 6e5350d.

File: tests/test_filter_rest.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+import unittest
+from queue import Queue
+
+from openfilter.filter_runtime.filters.rest import REST, RESTConfig
+from openfilter.filter_runtime.utils import adict
+
+logger = logging.getLogger(__name__)
+
+log_level = int(getattr(logging, (os.getenv('LOG_LEVEL') or 'CRITICAL').upper()))
+
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply tests/test_filter_rest.py]: Removed the unused `log_level` variable. See 6e5350d.

File: openfilter/filter_runtime/filters/rest.py [hunk 4]
```diff
@@ -30,6 +30,9 @@ class Endpoint(adict):
     declared_fps:  float | None
     resource_path: str | None   # Path to a resource directory that is accessible for local file loading.
 
+    auth_token:   str | None    # When set, require ?token= or Authorization: Bearer on all requests
+    cors_origins: str | None    # Comma-separated allowed CORS origins (default: '*')
+
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/rest.py:36]: Added `auth_token` and `cors_origins` to the REST docstring (including env var references). See 6e5350d.

File: openfilter/filter_runtime/filters/webvis.py [hunk 5]
```diff
@@ -19,6 +19,8 @@ class WebvisConfig(FilterConfig):
     port: int | None
     enable_json: bool = False
     sleep_interval: float = 1.0
+    auth_token: str | None = None
+    cors_origins: str | None = None
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/webvis.py:23]: Added `auth_token` and `cors_origins` to the Webvis docstring (including env var references). See 6e5350d.

File: openfilter/filter_runtime/filters/http_security.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+from typing import Optional
+
+logger = logging.getLogger(__name__)
+
+
+def parse_cors_origins(cors_origins: Optional[str]) -> list[str]:
+    """Parse a comma-separated CORS origins string. Returns ['*'] if empty/None."""
+    if not cors_origins or not cors_origins.strip():
+        return ['*']
+    return [origin.strip() for origin in cors_origins.split(',') if origin.strip()]
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py]: Good catch — `parse_cors_origins(",")` was returning `[]` instead of `["*"]`. Fixed to fall back to wildcard when the parsed list is empty after filtering. Added test. See ffd8aa5.

File: openfilter/filter_runtime/filters/http_security.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        allow_origins=origins,
+        allow_credentials=True,
+        allow_methods=['*'],
+        allow_headers=['*'],
+    )
+
+    token = auth_token.strip() if auth_token else None
+    if token:
+        add_token_auth_middleware(app, token)
+        logger.info('Token authentication enabled')
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py]: You are right — reordered so token auth is added first, then CORS. Since `add_middleware()` prepends, CORS now runs outermost and applies `Access-Control-Allow-Origin` headers even on 401 responses. See ffd8aa5.

File: openfilter/filter_runtime/filters/rest.py [hunk 8]
```diff
@@ -76,6 +79,15 @@ class REST(Filter):
...
 
+        auth_token:
+            When set, all requests must include ``?token=<value>`` or ``Authorization: Bearer <value>``.
+            Returns 401 if missing or invalid. Also settable via ``FILTER_AUTH_TOKEN`` env var.
+
+        cors_origins:
+            Comma-separated list of allowed CORS origins. Defaults to ``'*'`` (allow all).
+            Example: ``'https://portal.plainsight.tech,https://localhost:5173'``.
+            Also settable via ``FILTER_CORS_ORIGINS`` env var.
+
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/rest.py:92]: Correct — REST does not read `FILTER_*` env vars (only Webvis does via its `normalize_config`). Removed the misleading env var references from the REST docstring. See ffd8aa5.

[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/rest.py:92]: Updated — instead of just removing the docs, added full `FILTER_*` env var support to `REST.normalize_config()` for all scalar config params: `host`, `port`, `base_path`, `declared_fps`, `resource_path`, `auth_token`, `cors_origins`. This matches the Webvis pattern. Restored the env var references in the docstring. See 0d7ffbc.

File: openfilter/filter_runtime/filters/http_security.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+            # Check query parameter first
+            request_token = request.query_params.get('token')
+
+            # Fall back to Authorization header
+            if not request_token:
+                auth_header = request.headers.get('authorization', '')
+                if auth_header.lower().startswith('bearer '):
+                    request_token = auth_header[7:].strip()
+
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py:67]: Good catch — changed to `if request_token is None:` so an empty `?token=` is treated as present (and fails auth) rather than falling back to the header. See f4daa97.

File: tests/test_filter_webvis.py [hunk 10]
```diff
@@ -53,5 +53,102 @@ def test_set_interval(self):
...
+        self.assertIsNone(normalized.cors_origins)
+
+    def test_auth_token_from_env_var(self):
+        os.environ['FILTER_AUTH_TOKEN'] = 'env-token'
+        try:
+            config = {'sources': 'tcp://localhost:5550'}
+            normalized = Webvis.normalize_config(config)
+            self.assertEqual(normalized.auth_token, 'env-token')
+        finally:
+            del os.environ['FILTER_AUTH_TOKEN']
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply tests/test_filter_webvis.py]: Fixed — replaced manual `os.environ` mutation with `patch.dict` for proper test isolation. See f4daa97.

File: tests/test_filter_webvis.py [hunk 11]
```diff
@@ -53,5 +53,102 @@ def test_set_interval(self):
...
+
+    def test_cors_origins_from_env_var(self):
+        os.environ['FILTER_CORS_ORIGINS'] = 'https://a.com,https://b.com'
+        try:
+            config = {'sources': 'tcp://localhost:5550'}
+            normalized = Webvis.normalize_config(config)
+            self.assertEqual(normalized.cors_origins, 'https://a.com,https://b.com')
+        finally:
+            del os.environ['FILTER_CORS_ORIGINS']
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply tests/test_filter_webvis.py]: Fixed — same `patch.dict` fix applied here. See f4daa97.

File: tests/test_filter_rest.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    def test_base_path_from_env(self):
+        config = self._normalize(FILTER_BASE_PATH='api/v1')
+        self.assertEqual(config.base_path, 'api/v1')
+
+    def test_defaults_to_none_when_unset(self):
+        for key in ('FILTER_AUTH_TOKEN', 'FILTER_CORS_ORIGINS', 'FILTER_PORT',
+                     'FILTER_DECLARED_FPS', 'FILTER_BASE_PATH', 'FILTER_RESOURCE_PATH'):
+            os.environ.pop(key, None)
+        config = self._normalize()
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply tests/test_filter_rest.py]: Fixed — wrapped the env var removals in `patch.dict` so original values are restored. See f4daa97.

**Review by shingonoide — DISMISSED** [review]
Good implementation with proper timing-safe token comparison (`secrets.compare_digest`) and solid test coverage (36+ tests). A few security concerns to address:

**HIGH - Token leakage via request URL:**
In `rest.py`, `'url': str(request.url)` stores the full URL including `?token=SECRET` in frame data, which propagates downstream through the pipeline. Any downstream filter that logs or stores frame data will leak the auth token. Additionally, uvicorn access logs at `info` level (the default) will log the full URL in plaintext. Suggestion: strip `token` from query params before storing the URL in frame data, and document the log exposure risk prominently.

**MEDIUM - allow_credentials=True with wildcard origins:**
`http_security.py` always sets `allow_credentials=True` regardless of whether origins are `['*']`. Starlette's CORSMiddleware reflects the request's Origin header when credentials=True with wildcard, effectively allowing any origin with credentials. Set `allow_credentials=bool(origins != ['*'])` or only True when specific origins are configured.

**MEDIUM - REST env mapping scope creep:**
The new env mapping block in `rest.py` adds `FILTER_HOST`, `FILTER_PORT`, `FILTER_BASE_PATH`, `FILTER_DECLARED_FPS`, `FILTER_RESOURCE_PATH` support that did not exist before. These are beyond the stated scope and `FILTER_HOST`/`FILTER_PORT` will be silently overridden by sources URL parsing. Suggest limiting to only `auth_token` and `cors_origins`.

**Positive:**
- Timing-safe comparison is correct and important
- CORS preflight bypass logic properly checks both Origin and Access-Control-Request-Method
- Middleware ordering is correct (auth first, then CORS)
- Whitespace/empty token handling prevents accidental lockout

**Review by shingonoide — COMMENTED** [review]
Updated review (corrected from previous REQUEST_CHANGES):

Good implementation with proper timing-safe token comparison and solid test coverage (36+ tests). Two suggestions:

**MEDIUM - allow_credentials=True with wildcard origins:**
`http_security.py` always sets `allow_credentials=True` regardless of whether origins are `['*']`. Starlette's CORSMiddleware reflects the request's Origin header when credentials=True with wildcard, effectively allowing any origin with credentials. Consider setting `allow_credentials=bool(origins != ['*'])`.

**MEDIUM - REST env mapping scope creep:**
The new env mapping block in `rest.py` adds `FILTER_HOST`, `FILTER_PORT`, `FILTER_BASE_PATH`, `FILTER_DECLARED_FPS`, `FILTER_RESOURCE_PATH` support that did not exist before. These are beyond the stated scope and some will be silently overridden by sources URL parsing. Suggest limiting to only `auth_token` and `cors_origins`.

Positive: timing-safe comparison, proper CORS preflight bypass, correct middleware ordering, good whitespace/empty token handling.

Looks good from our side with these suggestions addressed.

**Comment by lucasmundim** [comment]
Thanks for the careful review @shingonoide! Both MEDIUM findings addressed in 9e9740f:

**1. allow_credentials with wildcard origins — fixed**

You're right that Starlette's CORSMiddleware reflects the Origin header when `allow_credentials=True` + `allow_origins=["*"]`, which effectively allows any origin with credentials. Changed to:

```python
allow_credentials = origins != ["*"]
```

- Wildcard origins → credentials disabled → spec-compliant `Access-Control-Allow-Origin: *`
- Specific origins → credentials enabled → cookies/Authorization headers work for those origins

Added explicit tests covering both cases.

**2. REST env mapping scope — partially addressed**

After discussing internally: the scope creep concern is half right. Adding env vars for *all* scalar params was an explicit ask from our team for deployment-time overrides. However, your technical point about silent override is valid: `host`/`port`/`base_path` come from the `sources` URL and would be overwritten by URL parsing after the env var assignment, so setting `FILTER_HOST` while using `sources` is a silent no-op.

Split the difference: dropped `host`/`port`/`base_path` from the env mapping and kept `declared_fps`, `resource_path`, `auth_token`, `cors_origins` (all non-overlapping with `sources` URL parsing). Restored the env var docs accordingly.

**Separate topic — REST platform exposure**

While investigating your review I realized the current plainsight-api export logic (`isServiceNode` + `IsTerminalNode` gate in `internal/filterpipeline/export/k8s.go`) only exposes leaf nodes, so REST (an Input filter = entry node) wouldn't be exposed via the Gateway even if it had a `port` parameter. This is a real gap if we want to allow external event injection into running pipelines, but it's out of scope for this PR — will open a follow-up for plainsight-api + pipeline editor + portal changes.

**Comment by shingonoide** [comment]
Both fixes look correct:

1. **CORS credentials** — good catch on the Starlette workaround. Disabling `allow_credentials` with wildcard origins is the right call per spec.
2. **REST env var cleanup** — removing duplicate config sources prevents silent override bugs.

LGTM, approving.

**Review by shingonoide — DISMISSED** [review]
Fixes verified. Clean implementation.

File: openfilter/filter_runtime/filters/http_security.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            # Check query parameter first
+            request_token = request.query_params.get('token')
+
+            # Fall back to Authorization header only if query param was absent
+            if request_token is None:
+                auth_header = request.headers.get('authorization', '')
+                if auth_header.lower().startswith('bearer '):
+                    request_token = auth_header[7:].strip()
+
+            if not request_token or not secrets.compare_digest(request_token, token):
```
Comment (shingonoide) [inline openfilter/filter_runtime/filters/http_security.py]: **Should-fix (silent-500 footgun, not blocking).** `secrets.compare_digest` raises `TypeError: comparing strings with non-ASCII characters is not supported` when either argument contains non-ASCII chars. There's no try/except here, so a token with any UTF-8 byte (smart quote, em-dash, `café`, etc. — common when operators copy-paste from password managers or docs) produces an unhandled 500 on every request — not a 401, no log hint about the cause. Fail-closed so it's not a bypass, but operators will waste hours debugging and likely shorten the token to get past it.

One-line fix:
```python
if not request_token or not secrets.compare_digest(request_token.encode('utf-8'), token.encode('utf-8')):
```
Or validate ASCII-only at config time with a clear `ValueError`. Reproduces deterministically with `configure_http_security(app, auth_token='café'); GET /x?token=ascii → 500`.

File: openfilter/filter_runtime/filters/http_security.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    app.add_middleware(
+        CORSMiddleware,
+        allow_origins=origins,
+        allow_credentials=allow_credentials,
+        allow_methods=['*'],
+        allow_headers=['*'],
+    )
+
+    if token and origins == ['*']:
+        logger.warning('Auth is enabled but CORS allows all origins — consider setting cors_origins')
```
Comment (shingonoide) [inline openfilter/filter_runtime/filters/http_security.py:120]: NIT: nice warning on token+wildcard. Consider also warning (or INFO-logging) when `auth_token` is unset AND the filter will bind to `0.0.0.0` — the "serving without authentication" line is easy to miss in a busy log stream. Not for this PR.

File: openfilter/filter_runtime/filters/http_security.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+from typing import Optional
+
+logger = logging.getLogger(__name__)
+
+
+def parse_cors_origins(cors_origins: Optional[str]) -> list[str]:
+    """Parse a comma-separated CORS origins string. Returns ['*'] if empty/None."""
+    if not cors_origins or not cors_origins.strip():
+        return ['*']
+    origins = [origin.strip() for origin in cors_origins.split(',') if origin.strip()]
```
Comment (shingonoide) [inline openfilter/filter_runtime/filters/http_security.py:32]: NIT: `parse_cors_origins` doesn't validate that each entry looks like a proper browser Origin (scheme + host + optional port, no trailing slash/path). A config typo like `example.com` / `https://example.com/` / `htpts://…` is silently accepted and then never matches any real browser Origin header. A `logger.warning` for malformed entries would help ops spot misconfig before the first request fails. Not blocking.

**Review by shingonoide — DISMISSED** [review]
Follow-up on my dismissed review from 2026-04-10: all three concerns (`allow_credentials=True` with wildcard origins, REST env mapping scope creep, token-leakage-via-frame-data) were addressed and are verified on `5a4160d`.

Re-read the full diff at this HEAD, including the v0.1.30 version bump. Core implementation is clean and the security primitives are right:

- `secrets.compare_digest` for timing-safe token check
- OPTIONS preflight bypass properly scoped to `Origin` + `Access-Control-Request-Method` (plain OPTIONS without CORS headers still 401 — verified by `test_plain_options_without_cors_headers_requires_auth`)
- Middleware ordering (CORS outermost, auth inner) so 401 responses still carry CORS headers
- `allow_credentials` correctly disabled on wildcard origins (Starlette would otherwise reflect Origin — classic CORS bypass)
- Empty `?token=` no longer falls back to `Authorization` header (the exact bug flagged in the dismissed CR)
- REST env mapping scoped to params that don't clash with sources URL parsing
- No token value written to logs anywhere; only `"Token authentication enabled"` info line

Test coverage is adversarial (69 new test methods across 3 files, exercising wrong/empty token 401s, preflight scope tightness, unknown-origin rejection, wildcard-plus-credentials invariants, query-over-header precedence). CI: `security-scan` / DCO / cascade-dry-run green; unit-tests pending at review time.

Approving. Two things surfaced on a deeper DA pass that I'd like flagged before merge — neither blocks approval, both are defensively-nice to land:

**1. Non-ASCII `auth_token` triggers an unhandled 500 (see inline on `http_security.py:55`).** `secrets.compare_digest` raises `TypeError` on non-ASCII chars. Operator pastes a token with curly quotes / em-dash / any UTF-8 byte — every request 500s with no 401, no log hint. Fail-closed (not a bypass) but a real footgun. One-line fix: encode to bytes first.

**2. REST `request_to_frame` still captures `Authorization` header + `?token=` query param into frame data.** At `rest.py:186-187`, `multi_items_to_dict(request.headers.items())` and `multi_items_to_dict(request.query_params.multi_items())` are copied verbatim into `data['http']`. Similarly `str(request.url)` at `:185` includes the token. Previously flagged and withdrawn last round because those fields weren't secret-bearing. **But this PR is what makes the `Authorization` header + `?token=` sensitive** — once an `auth_token` is configured, those values now flow downstream through every filter (logs, metrics, traces, OpenLineage events, any persisted frame dump). No redaction. The symmetric fix belongs with the PR that introduces the secret concept.

Minimal patch for #2: in `request_to_frame`, strip `authorization` (case-insensitive) from the captured headers dict and `token` from query_params / the stored URL string before assembling `data['http']`. One-liner each. (Those lines are outside the current diff so no inline there — but the fix is small enough to roll in.)

Also three smaller NITs inline — optional.

One tiny correction on `RELEASE.md` — *"fully backwards compatible"* is slightly overstated: default wildcard-CORS credentials behavior changed (`allow_credentials=True` → `False` on `['*']`). That change is a security improvement (and what we asked for) but not strictly backwards-compatible for any caller sending cookies over CORS with the default wildcard setup. Worth a one-line acknowledgment in the release notes.

LGTM.

File: openfilter/filter_runtime/filters/http_security.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+        app: The FastAPI application to configure.
+        auth_token: When set, require this token on all requests.
+        cors_origins: Comma-separated allowed origins. Defaults to '*'.
+    """
+    from fastapi.middleware.cors import CORSMiddleware
+
+    origins = parse_cors_origins(cors_origins)
+
+    # Middleware order matters: add_middleware() prepends, so the last added
+    # runs outermost. Add token auth first, then CORS, so CORS wraps auth
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py:99]: Addressed in 928dba8: `parse_cors_origins()` now normalizes `*` mixed with specific origins to just `["*"]` — mixing the two is invalid per the CORS spec. Added a test for the edge case.

File: openfilter/filter_runtime/filters/http_security.py (hunk 13 above)
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py]: Addressed in b422d98: `secrets.compare_digest` now compares UTF-8 encoded bytes, preventing the `TypeError` on non-ASCII tokens.

File: openfilter/filter_runtime/filters/http_security.py (hunk 14 above)
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py:120]: Good idea — noted for a follow-up. The current `"serving without authentication"` log line is at INFO level which should be visible enough for now.

File: openfilter/filter_runtime/filters/http_security.py (hunk 15 above)
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py:32]: Addressed in b422d98: `parse_cors_origins()` now logs a warning for origins missing `http://`/`https://` scheme or with a trailing slash.

**Review by shingonoide — DISMISSED** [review]
Re-approve at `b422d98` — our APPROVE at `5a4160d` was auto-dismissed by branch-protection on `main` when these two fix commits landed.

Both follow-ups from the dismissed APPROVE address the flagged items cleanly:
- **`b422d98`:** `secrets.compare_digest` now compares UTF-8-encoded bytes on both sides (S1 — the non-ASCII-token 500 is gone). Constant-time property preserved.
- **`b422d98`:** `parse_cors_origins` warns on origins missing `http://` / `https://` scheme or with trailing slash (N3).
- **`928dba8`:** `'*'` mixed with specific origins collapses to `['*']` per CORS spec, with a test covering both leading- and middle-wildcard cases. Nice catch beyond what I flagged — this also strengthens the `allow_credentials = origins != ['*']` guard by preventing credentials from sneaking through a mixed-wildcard list.

All 9 security invariants from the prior pass still hold: constant-time cmp, strict preflight bypass (Origin + ACRM required), CORS-outer / auth-inner ordering, empty-`?token=` no-fallback, no token logged.

One NIT for a future PR: adding an explicit non-ASCII token test case (e.g. `auth_token='sëcret🔑'`) would lock the encoding fix in regression — not blocking, just defensive.

S2 (REST `request_to_frame` captures `Authorization` header + `?token=` query param into frame data — noted in the body of the prior APPROVE) is still open; re-raising as a non-blocking follow-up if you want to land it separately. Worth flagging because once an `auth_token` is configured those values are now secrets riding downstream through every filter.

Everything else from the prior APPROVE stands. LGTM.

File: openfilter/filter_runtime/filters/webvis.py [hunk 17]
```diff
@@ -37,24 +39,27 @@ class Webvis(Filter):
 
         port:
             Default 8000.
+
+        auth_token:
+            When set, all requests must include ``?token=<value>`` or ``Authorization: Bearer <value>``.
+            Returns 401 if missing or invalid. Also settable via ``FILTER_AUTH_TOKEN`` env var.
+
+        cors_origins:
+            Comma-separated list of allowed CORS origins. Defaults to ``'*'`` (allow all).
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/webvis.py]: Addressed in 2eb96cf: webvis docstring now notes that wildcard CORS disables `Access-Control-Allow-Credentials` and specific origins are needed to enable it.

File: openfilter/filter_runtime/filters/rest.py [hunk 18]
```diff
@@ -76,6 +79,15 @@ class REST(Filter):
             If passing images via a filename it will be relative to this path which MUST be provided in order to pass
             images in this manner.
 
+        auth_token:
+            When set, all requests must include ``?token=<value>`` or ``Authorization: Bearer <value>``.
+            Returns 401 if missing or invalid. Also settable via ``FILTER_AUTH_TOKEN`` env var.
+
+        cors_origins:
+            Comma-separated list of allowed CORS origins. Defaults to ``'*'`` (allow all).
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/rest.py]: Addressed in 2eb96cf: same update applied to the REST filter docstring.

File: openfilter/filter_runtime/filters/http_security.py [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+            # Check query parameter first
+            request_token = request.query_params.get('token')
+
+            # Fall back to Authorization header only if query param was absent
+            if request_token is None:
+                auth_header = request.headers.get('authorization', '')
+                if auth_header.lower().startswith('bearer '):
+                    request_token = auth_header[7:].strip()
+
+            if not request_token or not secrets.compare_digest(request_token.encode('utf-8'), token.encode('utf-8')):
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply openfilter/filter_runtime/filters/http_security.py]: Addressed in 2eb96cf: token bytes are now precomputed once at middleware init (`token_bytes = token.encode("utf-8")`). Only the request token is encoded per-request.

**Review by shingonoide — DISMISSED** [review]
Re-approve at `eb4a5c7` — branch-protection on `main` auto-dismissed the previous APPROVE (`b422d98`) when the polish commit landed.

Two follow-ups since the last APPROVE, both additive:
- **`2eb96cf`:** `token_bytes = token.encode('utf-8')` now precomputed once at middleware init instead of per-request (addresses the DA nit). `compare_digest(request_token.encode('utf-8'), token_bytes)` remains constant-time — just one less per-request allocation. Plus webvis / REST docstrings now explicitly document that wildcard CORS disables `Access-Control-Allow-Credentials` and that specific origins are needed to enable it.
- **`eb4a5c7`:** adds non-ASCII token regression tests (`'sëcret'` acceptance + rejection). Exactly the NIT I asked for last round — thanks.

9 security invariants, N3 fix, and wildcard-mix collapse all preserved. No behavior change on the happy path.

S2 (REST `request_to_frame` captures `Authorization` header + `?token=` query param into frame data — body note from the earlier APPROVE) is still open; flagging again as a non-blocking follow-up. Once an `auth_token` is configured those values are secret-bearing and ride downstream through every filter.

Everything else stands. LGTM.

**Comment by lucasmundim** [comment]
Addressed S2 from @shingonoide's review in 074d053: `request_to_frame` now strips `Authorization` header, `token` query param, and rebuilds the URL without the token before writing to frame data. Test added verifying credentials are stripped while other params/headers are preserved.

Also added non-ASCII token regression tests (`sëcret🔑`) in eb4a5c7 as suggested.

**Review by shingonoide — APPROVED** [review]
Re-approve at `074d053` — S2 closed cleanly. Prior APPROVE at `eb4a5c7` auto-dismissed by branch protection on `main` when this fix commit landed.

`074d053` strips auth credentials from REST frame data before it propagates downstream: `Authorization` header removed (case-insensitive), `token` query param removed (case-insensitive), URL rebuilt without the token. Other headers/params preserved. Test `test_auth_credentials_stripped_from_frame_data` asserts 200 + token absent from `headers`/`query_params`/`url` + non-auth values (`other=keep`, `X-Custom`) preserved. That closes the only open item from the earlier pass.

Two micro-nits worth a future follow-up (not blocking):
1. URL rebuild uses raw `f'{k}={v}'` with no `urllib.parse.quote` — if a preserved query value contains `&`, `=`, `#`, space, or unicode, the stored `url` string is cosmetically malformed. The `query_params` dict stays authoritative for downstream consumers, so impact is purely display.
2. `Cookie` / `Proxy-Authorization` headers aren't stripped. REST doesn't advertise cookie auth and proxy-auth is unlikely in-cluster, so scope here is fine — just noting for when filters grow auth surfaces.

All prior security invariants preserved: constant-time token compare, precomputed `token_bytes`, preflight bypass scope, middleware ordering, `allow_credentials` off on wildcard, wildcard-mix collapse, N3 malformed-origin warn, UTF-8 encode on both sides of `compare_digest`. Non-ASCII regression tests from `eb4a5c7` lock the UTF-8 fix in place.

LGTM.

**Review by shingonoide — DISMISSED** [review]
APPROVE @ `074d0534`

## Since v4 (`b422d98c`)
- `074d0534` REST frame-data credential strip — **closes S2** (token/Authorization leakage into downstream filters).
- `2eb96cf3` precompute `token_bytes` once; docstring clarifies wildcard-CORS disables credentials. No behavior change.
- `eb4a5c74` non-ASCII token regression tests — pins S1 closed.

## Verified closed
- **S1** (non-ASCII auth_token → 500 via `compare_digest`): `openfilter/filter_runtime/filters/http_security.py:68` compares `bytes` to `bytes` (token precomputed at line 49). Regression tests at `tests/test_http_security.py:104-115` (`sëcret🔑`).
- **S2** (credential leakage into frame data): `openfilter/filter_runtime/filters/rest.py:183-189` strips `Authorization` header (case-insensitive), drops `token` query param, rebuilds URL without it. Direct assertion at `tests/test_filter_rest.py:126-149` — verifies header, query, and URL are all clean.

## Non-blocking observations
- `rest.py:189` rebuilds the query string with raw `f'{k}={v}'`, no `urllib.parse.quote`. Fine as a frame-data representation, but if any downstream filter re-issues the URL, special chars (`&`, `=`, spaces) in decoded values will corrupt the string. Follow-up ticket only if frame URLs ever become round-trippable.
- `Cookie` header isn't in the strip list. Not an auth surface today; revisit if cookie-based auth ever lands.
- Middleware ordering (`http_security.py:98-103`) is correct — CORS wraps auth, so 401 responses still carry `Access-Control-Allow-Origin`.

## Posture
APPROVE. Re-submitting at `074d0534` following auto-dismiss on the 3 new commits.

**Comment by lucasmundim** [comment]
Thanks for the thorough re-review @shingonoide. Ack on both non-blocking observations:

1. **URL rebuild without `urllib.parse.quote`** — agreed, `query_params` dict is authoritative for downstream consumers so the URL is informational only. Will address if frame URLs ever need to be round-trippable.
2. **`Cookie` / `Proxy-Authorization` headers** — noted. No auth surface today; will revisit if cookie-based auth is added.

Glad all security invariants are verified and S1/S2 are closed.

**Review by shingonoide — DISMISSED** [review]
Re-submitting following `dismiss_stale_reviews` on `main`. Net-new since v5 (`074d0534`):

1. **`rest.py` URL rebuild now uses `urlencode`** — correct handling of `&`, `=`, `+`, spaces, Unicode in query values. Removes a latent leak/corruption risk in the hand-rolled f-string join.
2. **Credential header strip widened** to `{authorization, cookie, proxy-authorization}` — good defense in depth; prevents session/proxy auth leaking into downstream frame data.
3. **Test coverage** for both behaviors plus CORS malformed-origin warnings.

Nits (feel free to ignore):
- `from urllib.parse import urlencode` could move to module-top imports.
- `test_url_encoding_of_special_chars_in_query` asserts only that keys are present; tightening to assert the actual encoded values (`hello%2Bworld`, `a%2526b`) would make the regression guard bite harder.

APPROVE.

**Comment by lucasmundim** [comment]
Addressed both nits from @shingonoide's review (4156415556) in 9fe7d85:

1. `from urllib.parse import urlencode` moved to module-level imports
2. `test_url_encoding_of_special_chars_in_query` now asserts exact encoded values (`hello+world`, `a%26b`) instead of just key presence

Also added a `wait-for-pypi` step to the release workflow in 21ea77a to fix the Docker image build race condition (PyPI index propagation delay).

File: .github/workflows/create-release.yaml [hunk 20]
```diff
@@ -262,9 +262,34 @@ jobs:
...
+        run: |
+          for i in $(seq 1 30); do
+            if pip install --dry-run "openfilter==$VERSION" 2>/dev/null; then
+              echo "✅ openfilter==$VERSION available on PyPI"
+              exit 0
+            fi
+            echo "Attempt $i/30: openfilter==$VERSION not yet indexed, waiting 10s..."
+            sleep 10
+          done
+          echo "❌ Timed out waiting for openfilter==$VERSION on PyPI" >&2
```
[the remark this answers is not in the document]
Comment (lucasmundim) [reply .github/workflows/create-release.yaml]: Addressed in d7c3662: now uses `needs.prepare-pypi-publish.outputs.package_name` instead of hardcoded `openfilter`.

**Review by shingonoide — APPROVED** [review]
Re-submitting APPROVE at `d7c36621`. Both v6 review nits are addressed cleanly:

- `urlencode` moved to module-top imports in `rest.py`.
- URL-encoding test now asserts exact encoded values (`msg=hello+world`, `tag=a%26b`) instead of just key presence.

The two CI commits (`21ea77a5`, `d7c36621`) are workflow-only — adding PyPI index-propagation polling before the Docker image build — and don't touch auth/CORS code. No regressions elsewhere in the auth module.

Thanks!
